docs/source/en/llm_optims.md

```diff
@@ -0,0 +1,324 @@
+
+> [!WARNING]
+> Currently, only [Command R](./model_doc/cohere), [Gemma](./model_doc/gemma) and [Llama](./model_doc/llama2) models support static kv-cache and torch.compile.
+
+For this example, let's load the [Gemma](https://hf.co/google/gemma-2b) model.
+
+```py
+from transformers import AutoTokenizer, AutoModelForCausalLM
+
+tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b")
+model = AutoModelForCausalLM.from_pretrained(
+    "google/gemma-2b", device_map="auto"
+)
+```
+
+There are two ways you can configure the model to use a static kv-cache.
+
+<hfoptions id="static-kv">
+<hfoption id="generation_config">
+
+Access the model's `generation_config` attribute and set the `cache_implementation` to "static".
+
+```py
+model.generation_config.cache_implementation = "static"
+```
+
+Call torch.compile on the model to compile the forward pass with the static kv-cache.
+
+```py
+compiled_model = torch.compile(model, mode="reduce-overhead", fullgraph=True)
+input_text = "The theory of special relativity states "
+input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")
+
+outputs = compiled_model.generate(**input_ids)
+tokenizer.batch_decode(outputs, skip_special_tokens=True)
+['The theory of special relativity states 1. The speed of light is constant in all inertial reference']
+```
+
+</hfoption>
+<hfoption id="setup_cache">
+
+> [!WARNING]
+> The `_setup_cache` method is an internal method and still under development, so keep in mind this will likely change in the future.
+
+The `_setup_cache` method doesn't support [`GenerationMixin.generate`] yet, so this method is a bit more involved. You'll need to write your own function to decode the next token given the current token and position and cache position of previously generated tokens.
+
+```py
+from transformers import LlamaTokenizer, LlamaForCausalLM, StaticCache, logging
+from transformers.testing_utils import CaptureLogger
+import torch
+
+prompts = [
+    "Simply put, the theory of relativity states that ",
+    "My favorite all time favorite condiment is ketchup.",
+]
+
+NUM_TOKENS_TO_GENERATE = 40
+torch_device = "cuda"
+
+tokenizer = LlamaTokenizer.from_pretrained("meta-llama/Llama-2-7b-hf", pad_token="</s>", padding_side="right")
+model = LlamaForCausalLM.from_pretrained("meta-llama/Llama-2-7b-hf", device_map="sequential")
+inputs = tokenizer(prompts, return_tensors="pt", padding=True).to(model.device)
+
+def decode_one_tokens(model, cur_token, input_pos, cache_position):
+    logits = model(
+        cur_token, position_ids=input_pos, cache_position=cache_position, return_dict=False, use_cache=True
+    )[0]
+    new_token = torch.argmax(logits[:, -1], dim=-1)[:, None]
+    return new_token
+```
+
+The are two important things you must do to enable static kv-cache and torch.compile with the `_setup_cache` method:
```

Comment:
I'm not sure about advertising using private methods - this means they're effectively public and need to have the same strict backwards compatibility handling of public methods

Reply:
I think it's ok with a warning that this is still experimental and its design will probably change in the future, so the user should be aware that some things may not work as expected, but in the meantime, you can still get excited/try this out? 

Would it help if I added a sentence about how it may not be backward compatible?

Reply:
That works for me!